ds/reactive/set_test.go

```diff
@@ -75,3 +77,283 @@ func TestSubtract(t *testing.T) {
+	s := newSet[int](1, 2, 3)
+	require.NotNil(t, s)
+	require.Equal(t, 3, s.Size())
+}
+
+func TestSet_Add(t *testing.T) {
+	s := newSet[int]()
+	added := s.Add(1)
+	require.True(t, added)
+	require.Equal(t, 1, s.Size())
+
+	// Test adding a duplicate element
+	added = s.Add(1)
+	require.False(t, added)
+	require.Equal(t, 1, s.Size())
+}
+
+func TestSet_AddAll(t *testing.T) {
+	s := newSet[int]()
+	otherSet := newSet[int](2, 3)
+	addedElements := s.AddAll(otherSet.ReadOnly())
+
+	require.Equal(t, 2, addedElements.Size())
+	require.Equal(t, 2, s.Size())
+
+	// Test adding duplicate elements
+	addedElements = s.AddAll(otherSet.ReadOnly())
+	require.Equal(t, 0, addedElements.Size())
+}
+
+func TestSet_Delete(t *testing.T) {
+	s := newSet[int](1, 2, 3)
+	deleted := s.Delete(2)
+	require.True(t, deleted)
+	require.Equal(t, 2, s.Size())
+
+	// Test deleting a non-existent element
+	deleted = s.Delete(4)
+	require.False(t, deleted)
+}
+
+func TestSet_DeleteAll(t *testing.T) {
+	s := newSet[int](1, 2, 3, 4)
+	otherSet := newSet[int](2, 3)
+	deletedElements := s.DeleteAll(otherSet.ReadOnly())
+
+	require.Equal(t, 2, deletedElements.Size())
+	require.Equal(t, 2, s.Size())
+}
+
+// Similar tests for Apply, Compute, Replace, Decode, ReadOnly
+// Note: Mocking or stubbing may be required for some methods
+
+// --- Tests for readableSet type ---
+
+func TestNewReadableSet(t *testing.T) {
+	r := newReadableSet[int](1, 2, 3)
+	require.NotNil(t, r)
+	require.Equal(t, 3, r.Size())
+}
+
+func TestNewDerivedSet(t *testing.T) {
+	d := newDerivedSet[int]()
+	require.NotNil(t, d)
+}
+
+// TestReadableSet_SubtractReactive tests the SubtractReactive method
+func TestReadableSet_SubtractReactive(t *testing.T) {
+	mainSet := newReadableSet[int](1, 2, 3)
+	subtractSet := newSet[int](2)
+
+	result := mainSet.SubtractReactive(subtractSet)
+
+	// Initially, 2 should be subtracted from mainSet
+	require.False(t, result.Has(2))
+	require.True(t, result.Has(1))
+	require.True(t, result.Has(3))
+
+	// Update subtractSet and check if result set is updated
+	subtractSet.Add(3)
+	require.False(t, result.Has(3)) // Now 3 should also be subtracted
+}
+
+// TestReadableSet_WithElements tests the WithElements method
+func TestReadableSet_WithElements(t *testing.T) {
+	rSet := newSet[int](1, 2, 3)
+
+	setupCalled := false
+	teardownCalled := false
+
+	teardown := rSet.WithElements(
+		func(element int) (teardown func()) {
+			setupCalled = true
+			return func() {
+				teardownCalled = true
+			}
+		},
+	)
+	defer teardown()
+
+	rSet.Add(4) // Trigger setup
+	require.True(t, setupCalled)
+
+	rSet.Delete(4) // Trigger teardown
+	require.True(t, teardownCalled)
+}
+
+// TestDerivedSet_inheritMutations tests the inheritMutations method
+func TestDerivedSet_inheritMutations(t *testing.T) {
+	dSet := newDerivedSet[int]()
+	mutations := ds.NewSetMutations[int]().WithAddedElements(newSet(1, 2))
+
+	// Apply mutations
+	appliedMutations := dSet.inheritMutations(mutations)
```

Comment:
I think it would be nice to have a duplicate in the inherited mutations?

Reply:
Added the element "2" twice: the set already excludes the repeated element though.